### PR TITLE
Example of MNIST with tqdm

### PR DESCRIPTION
**Pull request by rave78**
I have updated the basic example of MNIST with tqdm.  This example has two progress bars:  one progress bar counts the iterations whereas the other progress bar counts the epoch.

### CODE REVIEW

**Comment by vfdev-5** [comment]
@rave78 could you please fix flake8 problems in the modifications: https://travis-ci.org/pytorch/ignite/jobs/423061955

**Comment by vfdev-5** [comment]
Concerning the code itself, maybe it would be more interesting to replace `print` operations by tqdm.

**Comment by rave78** [comment]
Hi @vfdev-5,  
There is not print, only tqdm.write

**Comment by vfdev-5** [comment]
@rave78 sorry, I wanted to say to replace `mnist.py` example by `mnist-with-tqdm.py`, replace `print`s by tqdm...

**Comment by vfdev-5** [comment]
@rave78 tested your modification and gives me something like 
screen shot 2018-09-02 at 23 34 13

Could you please integrate "Loss: {:.2f}" into the progress bar and remove `tqdm.write` ?

**Comment by rave78** [comment]
@vfdev-5 sure, I can do those changes.  I will try to PR today

**Comment by vfdev-5** [comment]
@rave78 thanks, and please do not forget to add `tqdm` [here](https://github.com/pytorch/ignite/blob/master/.travis.yml#L28) as dependency

File: examples/mnist/mnist.py [hunk 1]
```diff
@@ -58,30 +60,64 @@ def run(train_batch_size, val_batch_size, epochs, lr, momentum, log_interval):
                                                      'nll': Loss(F.nll_loss)},
                                             device=device)
 
+    @trainer.on(Events.ITERATION_STARTED)
+    def log_training_loss_start(engine):
+        iter = (engine.state.iteration - 1) % len(train_loader) + 1
+        if iter == 1:
+            # engine.state.pdbar_iter = tqdm(
```
Comment (vfdev-5) [inline examples/mnist/mnist.py]: Maybe we could remove commented code

File: examples/mnist/mnist-with-tqdm.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
+from __future__ import print_function
```
Comment (vfdev-5) [inline examples/mnist/mnist-with-tqdm.py]: Maybe we can remove this file ?

File: examples/mnist/mnist.py (hunk 1 above)
Comment (rave78) [reply examples/mnist/mnist.py]: Sure, I will do it now (both changes)

File: examples/mnist/mnist-with-tqdm.py (hunk 2 above)
Comment (vfdev-5) [reply examples/mnist/mnist-with-tqdm.py]: @rave78 here I meant to remove the file `examples/mnist/mnist-with-tqdm.py` and not the line 
`from __future__ import print_function`. But you are right we do not need the same line in `mnist.py` neither

File: examples/mnist/mnist.py [hunk 3]
```diff
@@ -58,30 +59,61 @@ def run(train_batch_size, val_batch_size, epochs, lr, momentum, log_interval):
                                                      'nll': Loss(F.nll_loss)},
                                             device=device)
 
+    @trainer.on(Events.ITERATION_STARTED)
+    def log_training_loss_start(engine):
```
Comment (vfdev-5) [inline examples/mnist/mnist.py]: @rave78 I think it would be better have a single log function `log_training_loss` instead of two `log_training_loss_start` and `log_training_loss_end`. What do you think ?

Comment (rave78) [reply examples/mnist/mnist.py]: I think that makes total sense to merge both functions.  There is no difference starting the tqdm pbar at the ITERATION_STARTED or ITERATION_COMPLETED.

**Comment by vfdev-5** [comment]
@rave78 sorry about to request you modifications again and again, but if I look the [code difference : 
I have an impression that tqdm integration makes basic example with `print` more complex with tqdm. 
The modifications with tqdm specific handler function that you propose may perfectly be integrated as a separate tqdm handler in contrib, but here I think we should keep things as simple as with `print`. 
Let me define what we would like to keep and what we can replace : 
- please keep the same number of handlers. For example `pbar_epoch_start(engine)` should be avoided
- We can just use the progress bar for iterations (instead of a stack of lines as previously) and training/validation results can be printed with `tqdm.write` ?

What do you think ?

**Comment by rave78** [comment]
@vfdev-5 don't worry.  I am happy contributing with a few lines of code :).  Yes, it is possible to keep the original structure of mnist-example (in terms of handler functions).  Removing pdbar_epoch and using tqdm.write simplify a lot the code.  I will push those changes.

File: examples/mnist/mnist.py [hunk 4]
```diff
@@ -61,27 +62,39 @@ def run(train_batch_size, val_batch_size, epochs, lr, momentum, log_interval):
     @trainer.on(Events.ITERATION_COMPLETED)
     def log_training_loss(engine):
         iter = (engine.state.iteration - 1) % len(train_loader) + 1
-        if iter % log_interval == 0:
-            print("Epoch[{}] Iteration[{}/{}] Loss: {:.2f}"
-                  "".format(engine.state.epoch, iter, len(train_loader), engine.state.output))
+        desc = "ITERATION - loss: {:.2f}".format(engine.state.output)
+
+        if iter == 1:
```
Comment (vfdev-5) [inline examples/mnist/mnist.py]: @rave78 could we avoid if check like this ?

**Comment by vfdev-5** [comment]
@rave78 this looks much better. I left a comment on explicit `if iter == 0` checking

File: examples/mnist/mnist.py (hunk 4 above)
Comment (rave78) [reply examples/mnist/mnist.py]: Another option is to use `if not hasattr(engine.state, 'pdbar_iter')` instead of `if iter == 1`.  Note that `tqdm` is required for the first iteration for each epoch and `tqdm.close()` for the last iteration.

Comment (vfdev-5) [reply examples/mnist/mnist.py]: is it necessary to add `pdbar_iter` to `engine.state` ? Maybe you can define a `pbar = tqdm(...)` outside the handler like it is done for the [writer in tensorboard example](https://github.com/pytorch/ignite/blob/master/examples/mnist/mnist_with_tensorboardx.py#L80).

Comment (rave78) [reply examples/mnist/mnist.py]: I see.  Yes, it is possible reducing more the interaction between `tqdm` and the handler.

File: examples/mnist/mnist.py [hunk 5]
```diff
@@ -58,32 +59,47 @@ def run(train_batch_size, val_batch_size, epochs, lr, momentum, log_interval):
...
         metrics = evaluator.state.metrics
         avg_accuracy = metrics['accuracy']
         avg_nll = metrics['nll']
-        print("Validation Results - Epoch: {}  Avg accuracy: {:.2f} Avg loss: {:.2f}"
-              .format(engine.state.epoch, avg_accuracy, avg_nll))
+        tqdm.write(
+            "Validation Results - Epoch: {}  Avg accuracy: {:.2f} Avg loss: {:.2f}"
+            .format(engine.state.epoch, avg_accuracy, avg_nll))
+
+        pdbar.n = log_interval
```
Comment (vfdev-5) [inline examples/mnist/mnist.py]: @rave78 I wonder why we need the following code ?
```
pdbar.n = log_interval
pdbar.last_print_n = log_interval
pdbar.refresh()
```
And a nit: why `pdbar` and not just `pbar` ?

**Comment by vfdev-5** [comment]
I ran the code it looks good:
screen shot 2018-09-04 at 13 51 52

File: examples/mnist/mnist.py (hunk 5 above)
Comment (rave78) [reply examples/mnist/mnist.py]: @vfdev-5 the three lines of code set the starting state for the first iteration of the next epoch.  It is not a strong argument to use `pdbar` instead of `pbar`

File: examples/mnist/mnist.py [hunk 6]
```diff
@@ -58,32 +59,47 @@ def run(train_batch_size, val_batch_size, epochs, lr, momentum, log_interval):
                                                      'nll': Loss(F.nll_loss)},
                                             device=device)
 
+    desc = "ITERATION - loss: {:.2f}"
+    pdbar = tqdm(
+        initial=log_interval, leave=False, total=len(train_loader),
```
Comment (vfdev-5) [inline examples/mnist/mnist.py]: Shouldn't we this `initial=log_interval` and start from 0 ?

File: examples/mnist/mnist.py (hunk 5 above)
Comment (vfdev-5) [reply examples/mnist/mnist.py]: I would prefer to to compact this into 
```python
pdbar.n = pdbar.last_print_n = 0
```
starting from 0 instead of `log_interval`

File: examples/mnist/mnist.py (hunk 6 above)
Comment (rave78) [reply examples/mnist/mnist.py]: Yes, you are right.  I have set the initial value to zero.

File: examples/mnist/mnist.py (hunk 5 above)
Comment (rave78) [reply examples/mnist/mnist.py]: I have done both changes

**Comment by vfdev-5** [comment]
@rave78 LGTM thanks! I let @alykhantejani comment and merge if it is OK

**Comment by rave78** [comment]
@vfdev-5  sure :D.  Is there any other task that I can collaborate with?

**Comment by vfdev-5** [comment]
@rave78 thanks for the proposition ! Let's say we have two principal [projects](https://github.com/pytorch/ignite/projects) now : contrib module and portfolio examples.
Please take a look [here](https://github.com/pytorch/ignite/issues/64) or [here](https://github.com/pytorch/ignite/issues/202).
- a TensorboardLogger handler in contrib could be of interest. Something similar to #256. Main difficulty is to be enough generic and be simple :)
- port an example from pytorch/examples or something more "interesting" like segmentation example.
- we are also trying to find a solution to `create_supervised_*` methods to be able to configure `_prepare_batch`. Let me create an issue for that where I explain what we would like to have

**Comment by alykhantejani** [comment]
@rave78 thanks for the PR - sorry for the delay in merging!
